### PR TITLE
Changes for victory modals

### PR DESCRIPTION
**Pull request by shubhi1092**
- Includes changes for victory modal UX and its integration with all the systems (essentially PlayLevelView and PageIntroLevel)
- Enabled saving campaign slug to levels for ozaria so that we can easily know the campaign from level data
- Sending content data along with vue `completed` event in order to get name and type of content
- Moved buildLevelsList logic from PageUnitMap to vuex store so that it can be used in unit map as well as victory modals.

### CODE REVIEW

File: app/core/store/modules/unitMap.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      */
+    buildLevelsData: async ({ commit, rootGetters, dispatch }, campaignHandle, courseInstanceId) => { // campaignHandle necessary
+      let levels = {}
+      await dispatch('campaigns/fetch', campaignHandle, { root: true })
+      const campaignData = rootGetters['campaigns/getCampaignData'](campaignHandle)
+      if (!courseInstanceId) {
+        levels = campaignData.levels
+      } else {
+        try {
+          const courseInstance = await api.courseInstances.get({ courseInstanceID: courseInstanceId })
```
Comment (AndrewJakubowicz) [inline app/core/store/modules/unitMap.js]: Nitpick: Could this be moved as an addition to the courseInstances store module? Similar to how the campaign data is handled.

File: app/core/store/modules/unitMap.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+          const courseId = courseInstance.courseID
+          const classroomId = courseInstance.classroomID
+
+          // campaign snapshot of the levels
+          const existingCampaignLevels = _.cloneDeep(campaignData.levels)
```
Comment (AndrewJakubowicz) [inline app/core/store/modules/unitMap.js]: Should we import lodash explicitly with `import _ from 'lodash'`?

File: app/core/store/modules/unitMap.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+          // classroom snapshot of the levels for the course
+          const classroom = await api.classrooms.get({ classroomID: classroomId })
+          const classroomCourseLevels = _.find(classroom.courses, { _id: courseId }).levels
+
+          // get levels data for the levels in the classroom snapshot
+          const classroomCourseLevelsData = await api.classrooms.getCourseLevels({ classroomID: classroomId, courseID: courseId })
```
Comment (AndrewJakubowicz) [inline app/core/store/modules/unitMap.js]: How expensive is it to make 3 back and forth requests. When is it more efficient to make a new endpoint that wraps the back and forth exchange into one request?
Just a random thought, and requires no action.

File: app/core/store/modules/unitMap.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            }
+          }
+
+          levels = courseLevelsData
+        } catch (err) {
+          console.error('Error in building levels data', err)
+          // TODO: handle_error_ozaria
+        }
+      }
+      commit('setCurrentLevelsList', levels)
```
Comment (AndrewJakubowicz) [inline app/core/store/modules/unitMap.js]: Currently if there is an error in the courseInstance logic the levels object is still committed.
Is there a chance of any corrupt state being committed?

File: app/schemas/models/campaign.schema.coffee [hunk 5]
```diff
@@ -53,6 +53,7 @@ _.extend CampaignSchema.properties, {
         ]
     }
   }}
+  isOzaria: {type: 'boolean', description: 'Is this an ozaria campaign', default: true }
```
Comment (AndrewJakubowicz) [inline app/schemas/models/campaign.schema.coffee]: :clap:

File: ozaria/site/components/play/PageIntroLevel/index.vue [hunk 6]
```diff
@@ -69,163 +61,137 @@
...
-            await this.loadCampaignData()
-          } catch (err) {
-            console.error('Error in creating data for intro level', err)
-            // TODO handle_error_ozaria
-            noty({ text: 'Error in creating data for intro level', type: 'error', timeout: 2000 })
-            return
+        // Reading query params because this is rendered via backbone router and cannot be directly passed in as props
+        // They need to be in a specific order in the url to read and send them as props directy from backbone router, hence using query params here.
+        this.courseInstanceId = this.courseInstanceId || utils.getQueryVariable('course-instance')
+        this.codeLanguage = this.codeLanguage || utils.getQueryVariable('code-language')
```
Comment (AndrewJakubowicz) [inline ozaria/site/components/play/PageIntroLevel/index.vue]: Here we are checking the query variable `code-language`.

The naming switches between `code-language` and `codeLanguage`. Should only one be used?

File: ozaria/site/components/common/BaseModal.vue [hunk 7]
```diff
@@ -1,61 +1,66 @@
...
+    align-items: center
+    width: inherit
+    padding: 10px
 
-        transition: all .3s ease;
-    }
+  .ozaria-modal-header
+    font-weight: bold
+    font-size: 30px
+    font-family: $title-font-style
```
Comment (AndrewJakubowicz) [inline ozaria/site/components/common/BaseModal.vue]: Love the use of variables!

File: ozaria/site/components/common/ModalVictory.vue [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        let learningGoalsText
+        if (learningGoals) {
+          learningGoalsText = internationalizeConfig(learningGoals).body
+        }
+        return learningGoalsText
+      }
+    },
+    async mounted () {
+      // TODO Use new audio system post-august launch
+      Backbone.Mediator.publish('audio-player:play-sound', { trigger: 'victory' })
```
Comment (AndrewJakubowicz) [inline ozaria/site/components/common/ModalVictory.vue]: For my own understanding, if we change the trigger can we change the sound that plays? I think there will be different ozaria sounds that we'll need to incorporate.

File: ozaria/site/components/play/PageIntroLevel/index.vue [hunk 9]
```diff
@@ -69,163 +61,137 @@
...
+          return
+        }
+        // Assign first content in the sequence to this.currentContent
+        this.currentIndex = 0
+        this.currentContent = this.introContent[this.currentIndex]
+        this.setCurrentContentId(this.currentContent)
+        this.dataLoaded = true
+      },
+      onContentCompleted: async function (data) {
+        this.currentContentData = data || {}
```
Comment (AndrewJakubowicz) [inline ozaria/site/components/play/PageIntroLevel/index.vue]: I am a bit confused about this addition, and the cinematic emitting data in the completed event. Would you be able to elaborate what the purpose is?

What is the `this.currentContentData.contentType = this.currentContent.type` line doing?

File: ozaria/site/components/common/ModalVictory.vue [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+  import { mapActions, mapGetters } from 'vuex'
+  import { internationalizeConfig, getNextLevelForLevel, getNextLevelLink } from 'ozaria/site/common/ozariaUtils'
+  import utils from 'core/utils'
+
+  export default Vue.extend({
+    components: {
+      BaseModal
+    },
+    props: {
+      campaignHandle: {
```
Comment (AndrewJakubowicz) [inline ozaria/site/components/common/ModalVictory.vue]: Are any of these various props mutually exclusive. I.e. The modal only expects a campaignHandle or a currentLevel but not both?
Should that be handled in a console warning explicitly. Otherwise I assume all these props must be fulfilled.

File: ozaria/site/components/play/PageUnitMap/index.vue [hunk 11]
```diff
@@ -85,6 +86,7 @@
     methods: {
       ...mapActions({
         fetchCampaign: 'campaigns/fetch',
+        buildLevelsData: 'unitMap/buildLevelsData',
```
Comment (AndrewJakubowicz) [inline ozaria/site/components/play/PageUnitMap/index.vue]: Really like the refactor of this logic into Vuex.

File: ozaria/site/components/play/PageIntroLevel/index.vue [hunk 12]
```diff
@@ -69,163 +61,137 @@
...
           }
-        },
-        fetchNextLevel: async function () {
+          this.currentContentId = content.contentId[this.language]
+        } else {
+          this.currentContentId = content.contentId
+        }
+        // reload key is concatenated into child component's keys
+        // so that we can trigger their re-render with the same content id by incrementing reloadKey
+        // this is needed when `replay` is clicked on victory modal
```
Comment (AndrewJakubowicz) [inline ozaria/site/components/play/PageIntroLevel/index.vue]: Great comment =D

File: ozaria/site/components/play/PageIntroLevel/index.vue [hunk 13]
```diff
@@ -69,163 +61,137 @@
...
-      :key="currentContentId"
+      :key="currentContentId + `${reloadKey[currentContent.type]}`"
       :interactive-id-or-slug="currentContentId"
       :code-language="language"
       @completed="onContentCompleted"
     />
     <cinematics-component
       v-else-if="currentContent.type == 'cinematic'"
-      :key="currentContentId"
+      :key="currentContentId + `${reloadKey[currentContent.type]}`"
```
Comment (AndrewJakubowicz) [inline ozaria/site/components/play/PageIntroLevel/index.vue]: Could this be written like:

`${currentContentId}${reloadKey[currentContent.type]}`

Also does the reload number need to be tied to an object? Could it just be set statically on the data and incremented?

File: ozaria/site/components/play/PageIntroLevel/index.vue [hunk 14]
```diff
@@ -69,163 +61,137 @@
...
     />
     <avatar-selection-screen
       v-else-if="currentContent.type == 'avatarSelectionScreen'"
-      v-on:completed="onContentCompleted">
-    </avatar-selection-screen>
+      @completed="onContentCompleted"
+    />
+    <modal-victory
+      v-if="showVictoryModal"
+      :campaign-handle="introLevelData.campaign"
```
Comment (AndrewJakubowicz) [inline ozaria/site/components/play/PageIntroLevel/index.vue]: Could campaign specifically be driven from Vuex?

File: ozaria/site/components/play/PageUnitMap/index.vue [hunk 15]
```diff
@@ -127,57 +128,16 @@
...
-          const classroomCourseLevels = _.find(classroom.courses, { _id: courseId }).levels
+      async buildClassroomLevelMap () {
+        const courseInstance = await api.courseInstances.get({ courseInstanceID: this.courseInstanceId })
+        const courseId = this.courseId = courseInstance.courseID
+        const classroomId = courseInstance.classroomID
 
-          // get levels data for the levels in the classroom snapshot
-          const classroomCourseLevelsData = await api.classrooms.getCourseLevels({ classroomID: classroomId, courseID: courseId })
+        const classroom = this.classroom = await api.classrooms.get({ classroomID: classroomId })
+        const classroomCourseLevels = _.find(classroom.courses, { _id: courseId }).levels
```
Comment (AndrewJakubowicz) [inline ozaria/site/components/play/PageUnitMap/index.vue]: Nitpick: Import lodash explicitly?

File: ozaria/site/components/play/PageUnitMap/index.vue [hunk 16]
```diff
@@ -127,57 +128,16 @@
...
-            }
-            if (this.classroomLevelMap[original].first) {
-              courseLevelsData[original].first = this.classroomLevelMap[original].first
-            }
-          }
-
-          return courseLevelsData
-        } catch (err) {
-          return Promise.reject(err)
+        for (let level of classroomCourseLevels) {
```
Comment (AndrewJakubowicz) [inline ozaria/site/components/play/PageUnitMap/index.vue]: Nitpick: `const`

File: ozaria/site/styles/common/common.sass [hunk 17]
```diff
@@ -1,11 +1,26 @@
...
   font-weight: bold
   letter-spacing: 0.71px
-  line-height: 24px
\ No newline at end of file
+  line-height: 24px
+
+.ozaria-secondary-button
+  padding: 11px 19px
+  color: #1fbab4
+  font-family: "Open Sans"
```
Comment (AndrewJakubowicz) [inline ozaria/site/styles/common/common.sass]: Is this font family stored in a variable?

File: ozaria/site/views/play/level/PlayLevelView.js [hunk 18]
```diff
@@ -1447,10 +1411,12 @@ class PlayLevelView extends RootView {
       const showModalFn = () =>
         Backbone.Mediator.publish('level:show-victory', options)
       this.session.recordScores(this.world.scores, this.level)
-      if (this.level.get('replayable')) {
-        return this.session.increaseDifficulty(showModalFn)
-      } else {
-        return showModalFn()
+      if (this.level.get('ozariaType') !== 'capstone') {
+        if (this.level.get('replayable')) {
+          return this.session.increaseDifficulty(showModalFn)
```
Comment (AndrewJakubowicz) [inline ozaria/site/views/play/level/PlayLevelView.js]: This is a bit hard to follow. Could a comment be helpful?

File: ozaria/site/views/play/level/tome/SpellPaletteView.coffee [hunk 19]
```diff
@@ -206,7 +206,7 @@ module.exports = class SpellPaletteView extends CocoView
     @$el.toggleClass 'controls-disabled', not enabled
 
   onFrameChanged: (e) ->
-    return unless e.selectedThang?.id is @thang.id
+    return unless e.selectedThang?.id is @thang?.id
```
Comment (AndrewJakubowicz) [inline ozaria/site/views/play/level/tome/SpellPaletteView.coffee]: Reason for change?

File: test/ozaria/site/components/play/PageIntroLevel.spec.js [hunk 20]
```diff
@@ -51,25 +52,27 @@ describe('Intro level Page', () => {
...
-    expect(pageIntroLevelWrapper.contains(interactiveComponent)).toBe(true)
-  })
+  // TODO below tests need to be updated since they have been integrated with victory modals now
 
-  it('sets complete:true in the intro level session when all content completed', () => {
-    pageIntroLevelWrapper.find(cinematicComponent).vm.$emit('completed')
-    pageIntroLevelWrapper.find(interactiveComponent).vm.$emit('completed')
-    expect(api.levelSessions.update).toHaveBeenCalled()
-  })
+  // it('renders the content in correct sequence', () => {
```
Comment (AndrewJakubowicz) [inline test/ozaria/site/components/play/PageIntroLevel.spec.js]: Any specs that can be added via comments?

File: app/core/store/modules/unitMap.js (hunk 1 above)
Comment (shubhi1092) [reply app/core/store/modules/unitMap.js]: Yeah potentially, I will add a TODO comment for that here for later in the interest of time.

File: app/core/store/modules/unitMap.js (hunk 4 above)
Comment (shubhi1092) [reply app/core/store/modules/unitMap.js]: An empty object will be committed in that case because levels has been initialized as {} and is only updated at the end of all the logic that can error out.

File: ozaria/site/components/common/ModalVictory.vue (hunk 10 above)
Comment (shubhi1092) [reply ozaria/site/components/common/ModalVictory.vue]: It requires campaignHandle and currentLevel at the minimum, and those props are set as required. Do you think it should be validated in created/mounted method as well?

File: ozaria/site/components/common/ModalVictory.vue (hunk 8 above)
Comment (shubhi1092) [reply ozaria/site/components/common/ModalVictory.vue]: I am not sure I understand the question.

File: ozaria/site/components/play/PageIntroLevel/index.vue (hunk 6 above)
Comment (shubhi1092) [reply ozaria/site/components/play/PageIntroLevel/index.vue]: great pickup, will update

File: ozaria/site/components/play/PageIntroLevel/index.vue (hunk 9 above)
Comment (shubhi1092) [reply ozaria/site/components/play/PageIntroLevel/index.vue]: We need the data of the content to know its name for victory modals, and potentially other details in the future.
Cinematics, as well as other systems, are emitting that data with the completed event so that we do not need to call the api again to fetch data based on id/slug.
And then, `this.currentContentData.contentType = this.currentContent.type` adds the type cinematic/interactive/cutscene to that data because its needed for the victory modal.

File: ozaria/site/components/play/PageIntroLevel/index.vue (hunk 13 above)
Comment (shubhi1092) [reply ozaria/site/components/play/PageIntroLevel/index.vue]: I wanted to keep it differentiated for different contents like interactives/cinematics/cutscenes, technically using a single value would also work, but then any change would impact the keys for all the child components.

File: ozaria/site/components/play/PageIntroLevel/index.vue (hunk 14 above)
Comment (shubhi1092) [reply ozaria/site/components/play/PageIntroLevel/index.vue]: We just need the campaign id/slug that the level is part of(which is available in the level data and thus vuex not required), and later the complete data for the campaign is driven from vuex in the victory modal.

File: ozaria/site/styles/common/common.sass (hunk 17 above)
Comment (shubhi1092) [reply ozaria/site/styles/common/common.sass]: Yes, will update

File: ozaria/site/components/common/ModalVictory.vue (hunk 8 above)
Comment (AndrewJakubowicz) [reply ozaria/site/components/common/ModalVictory.vue]: Do we know the cost of changing the sound played for Ozaria?

File: ozaria/site/views/play/level/PlayLevelView.js (hunk 18 above)
Comment (shubhi1092) [reply ozaria/site/views/play/level/PlayLevelView.js]: I am not sure what this code does, but it triggered victory modal for capstone automatically, so I justed wanted to skip this for capstone.

File: ozaria/site/components/common/ModalVictory.vue (hunk 10 above)
Comment (AndrewJakubowicz) [reply ozaria/site/components/common/ModalVictory.vue]: Maybe a comment for now?

File: ozaria/site/views/play/level/tome/SpellPaletteView.coffee (hunk 19 above)
Comment (shubhi1092) [reply ozaria/site/views/play/level/tome/SpellPaletteView.coffee]: For some level, it was giving an error because thang was not defined, hence added this check

File: ozaria/site/components/play/PageIntroLevel/index.vue (hunk 13 above)
Comment (AndrewJakubowicz) [reply ozaria/site/components/play/PageIntroLevel/index.vue]: That's true. But only one content type is shown at a time. This is fine.

File: ozaria/site/components/play/PageIntroLevel/index.vue (hunk 14 above)
Comment (AndrewJakubowicz) [reply ozaria/site/components/play/PageIntroLevel/index.vue]: Could it be useful to store the current level in Vuex for other systems that need to know this info? Not priority.

File: test/ozaria/site/components/play/PageIntroLevel.spec.js (hunk 20 above)
Comment (shubhi1092) [reply test/ozaria/site/components/play/PageIntroLevel.spec.js]: I think we still want to test these scenarios, but would need to update how we test them. Cant think of additional tests as of now.

File: ozaria/site/components/common/ModalVictory.vue (hunk 8 above)
Comment (shubhi1092) [reply ozaria/site/components/common/ModalVictory.vue]: No as of now, we will have to consider that while implementing a new generic audio system

File: ozaria/site/components/play/PageIntroLevel/index.vue (hunk 13 above)
Comment (shubhi1092) [reply ozaria/site/components/play/PageIntroLevel/index.vue]: yeah thats true, but keeping it different would be cleaner to understand and less brittle.

File: ozaria/site/components/play/PageIntroLevel/index.vue (hunk 14 above)
Comment (shubhi1092) [reply ozaria/site/components/play/PageIntroLevel/index.vue]: Yeah it would definitely be useful to move levels to shared Vuex store at some point.

**Comment by shubhi1092** [comment]
Thanks @AndrewJakubowicz . 
Yes levels play for a really long time before triggering the victory modal, thats because they have a very large number of frames currently. The game team will soon update that, and then it will be shorter. Great pickup! 👍
